File: scripts/generate_rescue_figure.py
"""PGD-AT rescue / optimisation-stability figure.

Companion to the H2 AT ladder. Where the ladder (scripts/generate_at_ladder_figure.py)
reports the ORIGINAL unified PGD-AT protocol and marks collapsed points, this figure
isolates the follow-up question: *are those collapses intrinsic, or an artefact of an
under-stabilised optimiser?* It contrasts each collapsed point's original PGD-AT run
against a stronger-stabilisation "rescue" run (eps_warmup=8 / lr_warmup=5 / longer
schedule / LR halved / grad-clip), both under the same PGD-50+5restart strong eval.

The story the figure must carry:
  - OCT ResNet-152 collapses under the original protocol (robust@8 ~= 2.5%) but the
    rescue protocol fully recovers it (clean 97.8%, robust@8 89.2%) -> the collapse
    was optimisation, not capacity/task.
  - Chest ResNet-18 stays collapsed even under rescue -> collapses are heterogeneous;
    this one is recovered instead by switching method (TRADES/MART, see record 5.3c).

Two panels (Clean, PGD-8/255 robust), grouped bars original vs rescue per point,
collapsed bars drawn hollow with a red x at the base (same visual grammar as the
ladder figure). Reads the shared data file:
  figures/data/at_rescue.json   (written by scripts/extract_figure_data.py)

Usage:
    python scripts/generate_rescue_figure.py
"""

import json
import sys
from pathlib import Path

import matplotlib

matplotlib.use("Agg")

import matplotlib.pyplot as plt
import numpy as np
from matplotlib.lines import Line2D

# ── MANDATORY: editable SVG text (nature-figure skill) ──────────────────────
plt.rcParams["font.family"] = "sans-serif"
plt.rcParams["font.sans-serif"] = ["Arial", "DejaVu Sans", "Liberation Sans"]
plt.rcParams["svg.fonttype"] = "none"
plt.rcParams["pdf.fonttype"] = 42
plt.rcParams["font.size"] = 7
plt.rcParams["axes.spines.right"] = False
plt.rcParams["axes.spines.top"] = False
plt.rcParams["axes.linewidth"] = 0.8
plt.rcParams["legend.frameon"] = False

PROJECT_ROOT = Path(__file__).resolve().parent.parent

ORIG_COLOR = "#9AA6B2"    # muted grey-blue: the under-stabilised original run
RESCUE_COLOR = "#0F4D92"  # blue_main: the stronger-stabilisation rescue run
COLLAPSE_RED = "#B64342"

PANELS = [
    ("clean", "Clean accuracy"),
    ("robust8", "PGD-8/255 robust accuracy (full)"),
]


def load_rows():
    p = PROJECT_ROOT / "figures" / "data" / "at_rescue.json"
    data = json.load(open(p))
    rows = data["rows"]
    # stable, story-first order: recovered point first, then the still-collapsed one
    rows = sorted(rows, key=lambda r: (r["rescue_collapsed"], r["dataset"]))
    return data, rows


def _label(r):
    disp_model = {"resnet18": "ResNet-18", "resnet34": "ResNet-34",
                  "resnet50": "ResNet-50", "resnet101": "ResNet-101",
                  "resnet152": "ResNet-152"}.get(r["model"], r["model"])
    return f"{r['dataset_display']}\n{disp_model}"


def _bar(ax, xc, val, collapsed, color, width):
    """One bar; collapsed -> pale hollow bar + red x at the base."""
    ax.bar(xc, val, width=width,
           color=color if not collapsed else "white",
           edgecolor=color, linewidth=0.9,
           alpha=1.0 if not collapsed else 0.9, zorder=2)
    if collapsed:
        ax.plot(xc, 0.02, marker="x", ms=5, color=COLLAPSE_RED,
                mew=1.6, zorder=5, clip_on=False)


def panel(ax, rows, key, ylabel):
    x = np.arange(len(rows))
    w = 0.36
    for i, r in enumerate(rows):
        _bar(ax, i - w / 2, r[f"orig_{key}"], r["orig_collapsed"], ORIG_COLOR, w)
        _bar(ax, i + w / 2, r[f"rescue_{key}"], r["rescue_collapsed"], RESCUE_COLOR, w)
        # annotate the recovery delta on the robust panel for recovered points
        if key == "robust8" and not r["rescue_collapsed"] and r["orig_collapsed"]:
            top = r[f"rescue_{key}"]
            ax.annotate(f"+{(top - r['orig_robust8']) * 100:.0f} pts",
                        xy=(i + w / 2, top), xytext=(0, 3),
                        textcoords="offset points", ha="center", va="bottom",
                        fontsize=6.2, color=RESCUE_COLOR, fontweight="bold")
    ax.set_xticks(x)
    ax.set_xticklabels([_label(r) for r in rows], fontsize=6.4)
    ax.set_ylabel(ylabel)
    ax.set_ylim(0, 1.03)
    ax.set_yticks(np.arange(0, 1.01, 0.2))


def add_panel_label(ax, label):
    ax.text(-0.14, 1.03, label, transform=ax.transAxes,
            fontsize=10, fontweight="bold", ha="left", va="bottom")


def main():
    data, rows = load_rows()
    fig, axes = plt.subplots(1, 2, figsize=(6.4, 3.1))
    for ax, (key, ylabel) in zip(axes, PANELS):
        panel(ax, rows, key, ylabel)
    add_panel_label(axes[0], "a")
    add_panel_label(axes[1], "b")

    # shared legend below the panels
    handles = [
        Line2D([0], [0], marker="s", linestyle="none", markersize=7,
               mfc=ORIG_COLOR, mec=ORIG_COLOR, label="Original PGD-AT"),
        Line2D([0], [0], marker="s", linestyle="none", markersize=7,
               mfc=RESCUE_COLOR, mec=RESCUE_COLOR, label="PGD-AT rescue (stronger stabilisation)"),
        Line2D([0], [0], marker="x", linestyle="none", markersize=6,
               color=COLLAPSE_RED, mew=1.6, label="collapsed (trivial classifier)"),
    ]
    fig.legend(handles=handles, loc="lower center", ncol=3, fontsize=6.2,
               bbox_to_anchor=(0.5, -0.02))
    fig.tight_layout(rect=(0, 0.06, 1, 1), w_pad=2.0)

    out_dir = PROJECT_ROOT / "figures" / "at_ladder"
    out_dir.mkdir(parents=True, exist_ok=True)
    base = out_dir / "H2b_rescue_stability_py"
    for fmt in ("svg", "pdf", "png"):
        fig.savefig(f"{base}.{fmt}", dpi=600 if fmt == "png" else None,
                    bbox_inches="tight")
    plt.close(fig)
    print(f"saved {base}.svg / .pdf / .png")


if __name__ == "__main__":
    sys.exit(main())

File: scripts/generate_rescue_figure.R
# PGD-AT rescue / optimisation-stability figure -- R backend.
#
# ggplot2 + patchwork twin panels, same data & scientific claim as the Python
# version (scripts/generate_rescue_figure.py). Reads the shared data file:
#   figures/data/at_rescue.json   (written by scripts/extract_figure_data.py)
#
#   (a) Clean accuracy  : original PGD-AT vs rescue, per collapsed point
#   (b) PGD-8/255 robust: same; recovered points annotated with the gain
# Collapsed runs are drawn hollow (white fill + coloured outline) with a red x
# at the bar base, matching the ladder figure's visual grammar.
#
# Run:  Rscript scripts/generate_rescue_figure.R
# Needs: jsonlite, ggplot2, patchwork, svglite, ragg

suppressPackageStartupMessages({
  library(jsonlite)
  library(ggplot2)
  library(patchwork)
})

# Resolve project root from the script's own path (Rscript) or fall back to cwd.
script_path <- sub("^--file=", "",
                   grep("^--file=", commandArgs(FALSE), value = TRUE))
if (length(script_path) == 1 && nzchar(script_path)) {
  proj_root <- normalizePath(file.path(dirname(script_path), ".."), mustWork = FALSE)
} else {
  proj_root <- getwd()
}
if (!dir.exists(file.path(proj_root, "figures"))) proj_root <- getwd()

# ── Nature-style theme (shared with the ladder figure) ──────────────────────
theme_set(
  theme_classic(base_size = 6.5, base_family = "Arial") +
    theme(
      axis.line   = element_line(linewidth = 0.35, colour = "black"),
      axis.ticks  = element_line(linewidth = 0.35, colour = "black"),
      legend.title = element_blank(),
      legend.text = element_text(size = 5.8),
      legend.key.size = unit(3, "mm"),
      legend.position = "bottom",
      plot.tag    = element_text(size = 10, face = "bold"),
      panel.grid  = element_blank(),
      axis.text.x = element_text(size = 5.8)
    )
)

ORIG_COLOR   <- "#9AA6B2"  # under-stabilised original run
RESCUE_COLOR <- "#0F4D92"  # stronger-stabilisation rescue run
COLLAPSE_RED <- "#B64342"
PROTO_COLORS <- c("Original PGD-AT" = ORIG_COLOR,
                  "PGD-AT rescue"   = RESCUE_COLOR)
DISPLAY <- c(resnet18 = "ResNet-18", resnet34 = "ResNet-34", resnet50 = "ResNet-50",
             resnet101 = "ResNet-101", resnet152 = "ResNet-152")

# ── Data: reshape orig/rescue x clean/robust8 into a long frame ─────────────
rows <- fromJSON(file.path(proj_root, "figures", "data", "at_rescue.json"),
                 simplifyDataFrame = TRUE)$rows
# story-first order: recovered point (rescue not collapsed) first
rows <- rows[order(rows$rescue_collapsed, rows$dataset), ]
rows$point <- paste0(rows$dataset_display, "\n", DISPLAY[rows$model])
point_levels <- rows$point

long <- do.call(rbind, lapply(seq_len(nrow(rows)), function(i) {
  r <- rows[i, ]
  data.frame(
    point     = rep(r$point, 4),
    protocol  = c("Original PGD-AT", "PGD-AT rescue"),
    metric    = rep(c("clean", "robust8"), each = 2),
    value     = c(r$orig_clean, r$rescue_clean, r$orig_robust8, r$rescue_robust8),
    collapsed = c(r$orig_collapsed, r$rescue_collapsed,
                  r$orig_collapsed, r$rescue_collapsed),
    stringsAsFactors = FALSE
  )
}))
long$point    <- factor(long$point, levels = point_levels)
long$protocol <- factor(long$protocol, levels = names(PROTO_COLORS))

# recovery annotation: rescue succeeded where original had collapsed (robust panel)
ann <- do.call(rbind, lapply(seq_len(nrow(rows)), function(i) {
  r <- rows[i, ]
  if (!r$rescue_collapsed && r$orig_collapsed) {
    data.frame(point = r$point,
               value = r$rescue_robust8,
               label = sprintf("+%.0f pts", (r$rescue_robust8 - r$orig_robust8) * 100),
               stringsAsFactors = FALSE)
  } else NULL
}))

dodge <- position_dodge(width = 0.7)

panel <- function(metric_key, ylab, annotate = FALSE) {
  d <- subset(long, metric == metric_key)
  d$point <- factor(d$point, levels = point_levels)
  p <- ggplot(d, aes(x = point, y = value, fill = protocol)) +
    geom_col(aes(colour = protocol), position = dodge, width = 0.62,
             linewidth = 0.35) +
    # collapsed bars: overpaint white so only the coloured outline shows
    geom_col(data = subset(d, collapsed), fill = "white",
             aes(colour = protocol), position = dodge, width = 0.62,
             linewidth = 0.35, show.legend = FALSE) +
    # red x at the base of every collapsed bar (mapped to shape -> legend entry)
    geom_point(data = subset(d, collapsed),
               aes(y = 0.02, group = protocol,
                   shape = "collapsed (trivial classifier)"),
               position = dodge, colour = COLLAPSE_RED,
               size = 1.3, stroke = 0.65) +
    scale_fill_manual(values = PROTO_COLORS) +
    scale_colour_manual(values = PROTO_COLORS, guide = "none") +
    scale_shape_manual(name = NULL, values = c("collapsed (trivial classifier)" = 4)) +
    guides(shape = guide_legend(override.aes = list(colour = COLLAPSE_RED))) +
    coord_cartesian(ylim = c(0, 1.03)) +
    scale_y_continuous(breaks = seq(0, 1, 0.2)) +
    labs(x = NULL, y = ylab)
  if (annotate && !is.null(ann) && nrow(ann) > 0) {
    ad <- ann; ad$point <- factor(ad$point, levels = point_levels)
    p <- p + geom_text(data = ad,
                       aes(x = point, y = value, label = label),
                       inherit.aes = FALSE, colour = RESCUE_COLOR,
                       fontface = "bold", size = 2.1, vjust = -0.6,
                       nudge_x = 0.17)
  }
  p
}

pa <- panel("clean", "Clean accuracy") + labs(tag = "a") +
  theme(legend.position = "none")
pb <- panel("robust8", "PGD-8/255 robust accuracy (full)", annotate = TRUE) +
  labs(tag = "b")

fig <- pa + pb + plot_layout(ncol = 2, guides = "collect") &
  theme(legend.position = "bottom")

# ── Export (svglite / cairo_pdf / ragg, per skill R quick-start) ────────────
out_dir <- file.path(proj_root, "figures", "at_ladder")
dir.create(out_dir, showWarnings = FALSE, recursive = TRUE)
base <- file.path(out_dir, "H2b_rescue_stability_r")
w <- 160 / 25.4; h <- 82 / 25.4

svglite::svglite(paste0(base, ".svg"), width = w, height = h)
print(fig); dev.off()
grDevices::cairo_pdf(paste0(base, ".pdf"), width = w, height = h, family = "Arial")
print(fig); dev.off()
ragg::agg_png(paste0(base, ".png"), width = w, height = h, units = "in", res = 600)
print(fig); dev.off()

cat("saved", paste0(base, ".svg / .pdf / .png"), "\n")
